snapcraft/services/lifecycle.py

```diff
@@ -36,5 +68,49 @@ def setup(self) -> None:
             base=project.get_effective_base(),
             extra_build_snaps=project.get_extra_build_snaps(),
         )
-
         super().setup()
+
+    @overrides
+    def run(self, step_name: str | None, part_names: list[str] | None = None) -> None:
+        super().run(step_name, part_names)
+
+        enable_manifest = utils.strtobool(os.getenv("SNAPCRAFT_BUILD_INFO", "n"))
+        if enable_manifest:
+            self._generate_manifest()
+
+    def _generate_manifest(self) -> None:
+        """Create and populate the manifest file."""
+        emit.progress("Generating snap manifest...")
+        image_information = os.getenv("SNAPCRAFT_IMAGE_INFO", "{}")
+        primed_stage_packages: set[str] = set()
+
+        project = cast(models.Project, self._project)
+
+        parts = copy.deepcopy(project.parts)
+        for name, part in parts.items():
+            assets = self.get_pull_assets(part_name=name)
+            if assets:
+                part["stage-packages"] = assets.get("stage-packages", []) or []
+            for key in ("stage", "prime", "stage-packages", "build-packages"):
+                part.setdefault(key, [])
+
+            stage_packages = self.get_primed_stage_packages(part_name=name)
+            if stage_packages:
+                primed_stage_packages |= set(stage_packages)
+
+        host_arch = utils.get_host_architecture()
+        build_for = self._build_plan[0].build_for if self._build_plan else host_arch
+
+        manifest.write(
```

Comment:
writing of the manifest should happen in the pack service (write_metadata)

Reply:
Refactored to write the manifest along the rest of the metadata. Lifecycle-related data is still collected during the execution of the lifecycle service, as this information is not available elsewhere.